packages/propeller/src/FeedSource.mts

```diff
@@ -2,32 +2,93 @@ import { Batch, Sink } from "./Types.js"
 import { ICheckpoints } from "./Checkpoints.js"
 import { sleep } from "./Sleep.js"
 
-export class TailingFeedSource {
+class CheckpointWriter {
   constructor(
-    private readonly sourceId: string,
-    private readonly tailSleepIntervalMs: number,
     private readonly groupName: string,
+    private readonly trancheId: string,
+    private readonly intervalMs: number,
     private readonly checkpoints: ICheckpoints,
-    private readonly sink: Sink,
-    private readonly crawl: (
-      trancheId: string,
-      position: bigint,
-      signal: AbortSignal,
-    ) => AsyncIterable<Batch>,
-    private readonly establishOrigin?: (tranche: string) => Promise<bigint>,
+    private committedPos: bigint,
+    private validatedPos = committedPos,
   ) {}
 
-  async start(trancheId: string, signal: AbortSignal) {
-    let pos = await this.checkpoints.load(this.groupName, trancheId, this.establishOrigin)
+  async start(signal: AbortSignal) {
+    while (!signal.aborted) {
+      await sleep(this.intervalMs, signal)
+      if (this.committedPos !== this.validatedPos) {
+        await this.checkpoints.commit(this.groupName, this.trancheId, this.validatedPos)
+        this.validatedPos = this.committedPos
+      }
+    }
+  }
+
+  commit(pos: bigint) {
+    this.validatedPos = pos
+  }
+}
+
+export type TailingFeedSourceOptions = {
+  tailSleepIntervalMs: number
+  checkpointIntervalMs: number
+  groupName: string
+  checkpoints: ICheckpoints
+  sink: Sink
+  crawl: (trancheId: string, position: bigint, signal: AbortSignal) => AsyncIterable<Batch>
+  establishOrigin?: (tranche: string) => Promise<bigint>
```

Comment:
would be tempted to put this first too to reflect that it dictates where reading starts

this leaves the later args to group according to the logical flow

or you can put the sink first and have the reverse order too I guess